File: ios/telobike/telobike/TBMapViewController.m
//
//  TBSecondViewController.m
//  telobike
//
//  Created by Elad Ben-Israel on 9/23/13.
//  Copyright (c) 2013 Elad Ben-Israel. All rights reserved.
//

@import MapKit;
@import QuartzCore;

#import <SVGeocoder/SVGeocoder.h>

#import "TBStation.h"
#import "TBMapViewController.h"
#import "TBServer.h"
#import "NSObject+Binding.h"
#import "UIColor+Style.h"
#import "NSBundle+View.h"
#import "TBNavigationController.h"
#import "KMLParser.h"
#import "TBStationAnnotationView.h"
#import "TBAvailabilityView.h"
#import "TBFavorites.h"
#import "TBGoogleMapsRouting.h"
#import "TBFeedbackMailComposeViewController.h"
#import "UIViewController+GAI.h"
#import "TBPlacemarkAnnotation.h"

@interface TBMapViewController () <MKMapViewDelegate, MFMailComposeViewControllerDelegate>

@property (strong, nonatomic) IBOutlet MKMapView* mapView;
@property (strong, nonatomic) IBOutlet UIToolbar* bottomToolbar;
@property (strong, nonatomic) UIBarButtonItem* backButtonItem;

// station details
@property (strong, nonatomic) IBOutlet UIView* stationDetails;
@property (strong, nonatomic) IBOutlet TBAvailabilityView* stationAvailabilityView;
@property (strong, nonatomic) IBOutlet UILabel* availabilityLabel;
@property (strong, nonatomic) IBOutlet UIBarButtonItem* toggleStationFavoriteButton;
@property (strong, nonatomic) IBOutlet UIView* labelBackgroundView;
@property (strong, nonatomic) IBOutlet UIView* topFillerView;
@property (strong, nonatomic) IBOutlet NSLayoutConstraint* drawerTopConstraint;

@property (strong, nonatomic) TBServer* server;

// routes
@property (assign, nonatomic) BOOL routesVisible;
@property (strong, nonatomic) KMLParser* kmlParser;

@property (assign, nonatomic) BOOL regionChangingForSelection;

@end

@implementation TBMapViewController

#pragma mark - View controller events

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.backButtonItem = self.navigationItem.leftBarButtonItem;
    
    self.server = [TBServer instance];
    
    [self observeValueOfKeyPath:@"stations" object:self.server with:^(id new, id old) {
        [self reloadAnnotations];
    }];
    
    [self observeValueOfKeyPath:@"city" object:self.server with:^(id new, id old) {
        MKCoordinateRegion region;
        region.center = self.server.city.cityCenter.coordinate;
        region.span = MKCoordinateSpanMake(0.05, 0.05);
        [self.mapView setRegion:region animated:NO];
    }];
    
    // map view
    MKUserTrackingBarButtonItem* trackingBarButtonItem = [[MKUserTrackingBarButtonItem alloc] initWithMapView:self.mapView];
    [self.bottomToolbar setItems:[self.bottomToolbar.items arrayByAddingObject:trackingBarButtonItem]];
    self.mapView.showsUserLocation = YES;
    
    self.stationAvailabilityView.alignCenter = YES;
    
    [self closeDetailsAnimated:NO];
}

- (void)openDetailsAnimated:(BOOL)animated {
    self.drawerTopConstraint.constant = 0.0f;
    self.topFillerView.hidden = YES;
    
    if (animated) {
        [UIView animateWithDuration:0.5f
                              delay:0.0f
             usingSpringWithDamping:0.6f
              initialSpringVelocity:8.0f
                            options:0
                         animations:^{ [self.view layoutIfNeeded]; }
                         completion:nil];
    }
}

- (void)closeDetailsAnimated:(BOOL)animated {
    self.drawerTopConstraint.constant = -self.stationDetails.frame.size.height;
    self.topFillerView.hidden = NO;

    if (animated) {
        [UIView animateWithDuration:0.5f
                              delay:0.0f
             usingSpringWithDamping:0.6f
              initialSpringVelocity:-8.0f
                            options:0
                         animations:^{
                             [self.view layoutIfNeeded];
                         }
                         completion:nil];
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [[TBServer instance] reloadStations:nil];
    
    [self showOrHideRoutesOnMap];
    
    if (self.navigationController.viewControllers.count > 1) {
        self.navigationItem.leftBarButtonItem = self.backButtonItem;
    }
    else {
        self.navigationItem.leftBarButtonItem = nil;
    }
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self analyticsScreenDidAppear:@"map"];
}

#pragma mark - Navigation

- (IBAction)back:(id)sender {
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark - Annotations

- (void)showPlacemark:(SVPlacemark*)placemark {
    // delete any existing placemark annotations
    [self.mapView removeAnnotations:[self.mapView.annotations filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(id evaluatedObject, NSDictionary *bindings) {
        return [evaluatedObject isKindOfClass:[TBPlacemarkAnnotation class]];
    }]]];
    
    TBPlacemarkAnnotation* newAnnotation = [[TBPlacemarkAnnotation alloc] initWithPlacemark:placemark];
    [self.mapView addAnnotation:newAnnotation];
    [self.mapView selectAnnotation:newAnnotation animated:YES];
}

- (void)deselectAllAnnoations {
    for (id ann in self.mapView.annotations) {
        [self.mapView deselectAnnotation:ann animated:YES];
    }
}

- (void)selectAnnotation:(id<MKAnnotation>)annotation animated:(BOOL)animated {
    [self.mapView selectAnnotation:annotation animated:animated];
}

- (NSArray *)annoations {
    return self.mapView.annotations;
}

- (void)reloadAnnotations {
    // add stations that are not already defined as annoations
    NSMutableSet* newAnnotations = [NSMutableSet setWithArray:[TBServer instance].stations];
    [newAnnotations minusSet:[NSSet setWithArray:self.mapView.annotations]];
    [self.mapView addAnnotations:newAnnotations.allObjects];
}

- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id<MKAnnotation>)annotation {
    static NSString* stationID = @"station";
    static NSString* placemarkID = @"placemark";
    
    if ([annotation isKindOfClass:[TBPlacemarkAnnotation class]]) {
        MKAnnotationView* view = [mapView dequeueReusableAnnotationViewWithIdentifier:placemarkID];
        if (!view) {
            MKPinAnnotationView* v = [[MKPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:placemarkID];
            view = v;
            v.pinColor = MKPinAnnotationColorRed;
            v.animatesDrop = YES;
            v.canShowCallout = YES;
        }
        view.annotation = annotation;
        return view;
    }
    
    // only if this is a station annotation
    if ([annotation isKindOfClass:[TBStation class]]) {
        MKAnnotationView* view = [self.mapView dequeueReusableAnnotationViewWithIdentifier:stationID];
        if (!view) {
            view = [[TBStationAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:stationID];
        }
        view.annotation = annotation;
        return view;
    }
    
    return nil;
}

#pragma mark - Selection

- (void)updateTitle:(NSString*)title {
    self.navigationItem.title = title ? title : self.title;
}

- (void)mapView:(MKMapView *)mapView didDeselectAnnotationView:(MKAnnotationView *)view {
    [self closeDetailsAnimated:YES];
    [self updateTitle:nil];
}

- (void)mapView:(MKMapView *)mapView didSelectAnnotationView:(MKAnnotationView *)view {
    NSString* annotationTitle;
    MKCoordinateRegion annotationRegion;

    if ([view.annotation isKindOfClass:[TBStation class]]) {
        TBStation* selectedStation = (TBStation*)view.annotation;

        [self openDetailsAnimated:YES];
        
        MKCoordinateRegion region;
        region.span = MKCoordinateSpanMake(0.004, 0.004);
        region.center = selectedStation.coordinate;
        
        annotationRegion = region;
        annotationTitle = selectedStation.stationName;
        
        [self updateStationDetails];
    }
    
    if ([view.annotation isKindOfClass:[TBPlacemarkAnnotation class]]) {
        TBPlacemarkAnnotation* annoation = view.annotation;
        annotationTitle = annoation.placemark.formattedAddress;;
        annotationRegion = MKCoordinateRegionMakeWithDistance(annoation.coordinate, 1000.0f, 1000.0f);;
    }
    
    [self updateTitle:annotationTitle];
    self.regionChangingForSelection = YES;
    [self.mapView setRegion:annotationRegion animated:YES];
}

- (void)mapView:(MKMapView *)mapView regionWillChangeAnimated:(BOOL)animated {
    if (self.regionChangingForSelection) {
        return; // if region is changing for selection, do nothing
    }

    [self deselectAllAnnoations];
}

- (void)mapView:(MKMapView *)mapView regionDidChangeAnimated:(BOOL)animated {
    self.regionChangingForSelection = NO;
}

#pragma mark - Station details

- (TBStation*)openedStation {
    TBStation* selectedStation = self.mapView.selectedAnnotations[0];
    if (![selectedStation isKindOfClass:[TBStation class]]) {
        NSLog(@"WARNING: selected annotation is not a station");
        return nil;
    }

    return  selectedStation;
}

- (void)updateStationDetails {
    TBStation* station = self.openedStation;
    if (!station) {
        return;
    }
    
    self.stationAvailabilityView.station = station;
    
    NSString* labelText = nil;
    switch (station.state) {
        case StationFull:
            labelText = NSLocalizedString(@"No parking", nil);
            break;
            
        case StationEmpty:
            labelText = NSLocalizedString(@"No bicycles", nil);
            break;
            
        case StationMarginal:
            labelText = NSLocalizedString(@"Almost empty", nil);
            break;
            
        case StationMarginalFull:
            labelText = NSLocalizedString(@"Almost full", nil);
            break;
            
        case StationInactive:
            labelText = NSLocalizedString(@"Not operational", nil);
            break;
            
        case StationUnknown:
        case StationOK:
        default:
            break;
    }
    
    self.availabilityLabel.hidden = !labelText;
    self.availabilityLabel.text = labelText;
    self.availabilityLabel.textColor = station.indicatorColor;
    self.labelBackgroundView.hidden = self.availabilityLabel.hidden;
    
    // set favorite
    UIImage* favoriteButtonImage = station.isFavorite ?
        [UIImage imageNamed:@"station-favorite-selected"] :
        [UIImage imageNamed:@"station-favorite-unselected"];
    
    [self.toggleStationFavoriteButton setImage:favoriteButtonImage];
}

- (IBAction)toggleStationFavorite:(id)sender {
    [self.openedStation setFavorite:!self.openedStation.isFavorite];
    [self updateStationDetails];
}

- (IBAction)sendStationReport:(id)sender {
    TBFeedbackMailComposeViewController* vc = [[TBFeedbackMailComposeViewController alloc] initWithFeedbackOption:TBFeedbackActionSheetService];
    NSString* subject = [NSString stringWithFormat:NSLocalizedString(@"Problem in station %@", nil), self.openedStation.sid];
    vc.mailComposeDelegate = self;
    [vc setSubject:subject];
    
    NSString* body = [NSString stringWithFormat:NSLocalizedString(@"Please describe the problem:\n\n\n=====================\nStation ID: %@\nName: %@\nAddress: %@", nil),
                      self.openedStation.sid,
                      self.openedStation.stationName,
                      self.openedStation.address ? self.openedStation.address : NSLocalizedString(@"N/A", nil)];
    
    [vc setMessageBody:body isHTML:NO];
    
    [self presentViewController:vc animated:YES completion:nil];
}

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error {
    [self dismissViewControllerAnimated:YES completion:nil];
}


- (IBAction)navigateToStation:(id)sender {
    NSString* dest = [NSString stringWithFormat:@"%g,%g", self.openedStation.coordinate.latitude, self.openedStation.coordinate.longitude];
    if (![TBGoogleMapsRouting routeFromAddress:@"" toAddress:dest]) {
        [[[UIAlertView alloc] initWithTitle:NSLocalizedString(@"Google Maps is not installed", nil) message:nil delegate:nil cancelButtonTitle:NSLocalizedString(@"OK", nil) otherButtonTitles:nil] show];
    }
}

#pragma mark - My location

- (void)mapView:(MKMapView *)mapView didFailToLocateUserWithError:(NSError *)error {
    NSLog(@"ERROR: unable to determine location: %@", error);
}

#pragma mark - Routes

- (void)showOrHideRoutesOnMap {
    BOOL showRoutes = [[NSUserDefaults standardUserDefaults] boolForKey:@"show_bicycle_routes"];
    
    if (showRoutes && !self.routesVisible) {
        if (!self.kmlParser) {
            NSURL* url = [[NSBundle mainBundle] URLForResource:@"routes" withExtension:@"kml"];
            self.kmlParser = [[KMLParser alloc] initWithURL:url];
            [self.kmlParser parseKML];
        }
        
        [self.mapView addOverlays:self.kmlParser.overlays];
        self.routesVisible = YES;
        return;
    }
    
    if (!showRoutes && self.routesVisible) {
        [self.mapView removeOverlays:self.kmlParser.overlays];
        self.routesVisible = NO;
    }
}

- (MKOverlayRenderer *)mapView:(MKMapView *)mapView rendererForOverlay:(id<MKOverlay>)overlay {
    return [self.kmlParser rendererForOverlay:overlay];
}

//#pragma mark - Placemark search result
//
//- (void)setSelectedPlacemark:(TBPlacemarkAnnotation *)selectedPlacemark {
//
//    if (selectedPlacemark) {
//        [self deselectAllAnnoations];
//        self.selectedStation = nil;
//
//        // only remove previous placemark if there is a new one to select
//        // otherwise, we just want to deselect it.
//        [self.mapView removeAnnotation:_selectedPlacemark];
//        [self.mapView addAnnotation:selectedPlacemark];
//    }
//    
//    _selectedPlacemark = selectedPlacemark;
//    [self.mapView selectAnnotation:selectedPlacemark animated:YES];
//}

@end

File: ios/telobike/telobike/TBMapViewController.swift
//
//  TBMapViewController.swift
//  telobike
//
//  Created by Elad Ben-Israel on 11/16/14.
//  Copyright (c) 2014 Elad Ben-Israel. All rights reserved.
//

import UIKit
import MapKit
import MessageUI

class TBMapViewController: UIViewController, MKMapViewDelegate, MFMailComposeViewControllerDelegate {
    
    @IBOutlet private weak var mapView: MKMapView!
    @IBOutlet private weak var bottomToolbar: UIToolbar!
    private weak var backButtonItem: UIBarButtonItem!

    // station details
    @IBOutlet private weak var stationDetails: UIView!
    @IBOutlet private weak var stationAvailabilityView: TBAvailabilityView!
    @IBOutlet private weak var availabilityLabel: UILabel!
    @IBOutlet private weak var toggleStationFavoriteButton: UIBarButtonItem!
    @IBOutlet private weak var labelBackgroundView: UIView!
    @IBOutlet private weak var drawerTopConstraint: NSLayoutConstraint!

    // observers
    private var stationsObserver: TBObserver?
    private var cityObserver: TBObserver?
    private var currentStationObserver: TBObserver?
    
    private var server: TBServer { return TBServer.instance }

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.backButtonItem = self.navigationItem.leftBarButtonItem
        
        // observers
        
        self.stationsObserver = TBObserver.observerForObject(self.server, keyPath: "stationsUpdateTime") {
            self.reloadAnnotations()
        }
        
        self.cityObserver = TBObserver.observerForObject(self.server, keyPath: "cityUpdateTime") {
            if let center = self.server.city?.cityCenter.coordinate {
                self.setMapRegion(MKCoordinateRegion(center: center, span: MKCoordinateSpanMake(0.05, 0.05)),
                    animated: true,
                    withoutDeselection: true)
            }
        }

        // map view
        let trackingBarButtonItem = MKUserTrackingBarButtonItem(mapView: self.mapView)

        var bottomToolbarItems = self.bottomToolbar.items ?? [UIBarButtonItem]()
        bottomToolbarItems.append(trackingBarButtonItem)
        self.bottomToolbar.setItems(bottomToolbarItems, animated: true)
        self.mapView.showsUserLocation = true

        self.stationAvailabilityView.alignCenter = true
        self.updateStationDetails(nil, animated: false)
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        self.server.reloadStations(force: false)
        self.showOrHideRoutesOnMap()
        if self.navigationController?.viewControllers.count > 1 {
            self.navigationItem.leftBarButtonItem = self.backButtonItem
        }
        else {
            self.navigationItem.leftBarButtonItem = nil
        }
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        self.analyticsScreenDidAppear("map")
    }

    @IBAction func back(sender: AnyObject!) {
        self.navigationController?.popViewControllerAnimated(true)
    }
    
    // - MARK: Annotations
    
    func showPlacemark(placemark: SVPlacemark?) {
        // delete any existing placemark annotations
        let placemarks = self.mapView.annotations.filter { $0.isKindOfClass(TBPlacemarkAnnotation) }
        self.mapView.removeAnnotations(placemarks)

        if let pm = placemark {
            let newAnnotation = TBPlacemarkAnnotation(placemark: pm)
            self.mapView.addAnnotation(newAnnotation)
            self.mapView.selectAnnotation(newAnnotation, animated: true)
        }
    }
    
    func deselectAllAnnotations() {
        for ann in self.mapView.annotations {
            self.mapView.deselectAnnotation(ann as MKAnnotation, animated: true)
        }
    }
    
    func selectAnnotation(annotation: MKAnnotation!, animated: Bool) {
        self.mapView.selectAnnotation(annotation, animated: animated)
    }
    
    private func reloadAnnotations() {
        // add stations that are not already defined as annoations
        let newAnnotations = NSMutableSet(array: self.server.stations)
        newAnnotations.minusSet(NSSet(array: self.mapView.annotations) as Set<NSObject>)
        self.mapView.addAnnotations(newAnnotations.allObjects as! [MKAnnotation])
    }
    
    func mapView(mapView: MKMapView, viewForAnnotation annotation: MKAnnotation) -> MKAnnotationView? {
        if annotation.isKindOfClass(TBPlacemarkAnnotation) {
            let placemarkReuseID = "placemark"
            var view = mapView.dequeueReusableAnnotationViewWithIdentifier(placemarkReuseID)
            if view == nil {
                let v = MKPinAnnotationView(annotation: annotation, reuseIdentifier: placemarkReuseID)
                view = v
                v.pinColor = .Red
                v.animatesDrop = true
                v.canShowCallout = true
            }
            
            view!.annotation = annotation
            return view
        }
        
        if annotation.isKindOfClass(TBStation) {
            let stationReuseID = "station"
            var view = self.mapView.dequeueReusableAnnotationViewWithIdentifier(stationReuseID)
            if view == nil {
                view = TBStationAnnotationView(annotation: annotation, reuseIdentifier: stationReuseID)
            }
            view!.annotation = annotation
            return view
        }
        
        return nil
    }
    
    // - MARK: Selection
    
    private func updateTitle(title: String?) {
        self.navigationItem.title = title ?? self.title
    }
    
    func mapView(mapView: MKMapView, didDeselectAnnotationView view: MKAnnotationView) {
        self.updateStationDetails(nil, animated: true)
    }
    
    func mapView(mapView: MKMapView, didSelectAnnotationView view: MKAnnotationView) {
        var annotationRegion: MKCoordinateRegion?
        
        if let stationAnnotation = view.annotation as? TBStation {
            self.updateStationDetails(stationAnnotation, animated: true)
            annotationRegion = MKCoordinateRegion(
                center: stationAnnotation.coordinate,
                span: MKCoordinateSpan(latitudeDelta: 0.004, longitudeDelta: 0.004))
        }
        
        if let placemarkAnnotation = view.annotation as? TBPlacemarkAnnotation {
            self.updateTitle(placemarkAnnotation.placemark.formattedAddress)
            annotationRegion = MKCoordinateRegionMakeWithDistance(placemarkAnnotation.coordinate, 1000.0, 1000.0)
        }
        
        if annotationRegion != nil {
            self.setMapRegion(annotationRegion!, animated: true, withoutDeselection: true)
        }
    }
    
    // - MARK: Map Region Changes
    
    private var avoidDeselectionWhenChangingRegion = false
    
    func setMapRegion(region: MKCoordinateRegion, animated: Bool, withoutDeselection: Bool) {
        self.avoidDeselectionWhenChangingRegion = withoutDeselection
        self.mapView.setRegion(region, animated: animated)
    }
    
    func mapView(mapView: MKMapView, regionWillChangeAnimated animated: Bool) {
        // if region is changing for selection, do nothing
        if self.avoidDeselectionWhenChangingRegion { return; }
        self.deselectAllAnnotations()
    }
    
    func mapView(mapView: MKMapView, regionDidChangeAnimated animated: Bool) {
        self.avoidDeselectionWhenChangingRegion = false
    }
    
    // - MARK: Station Details
    
    private func openDetails(animated: Bool) {
        let f: () -> () = {
            self.drawerTopConstraint.constant = 0.0
            self.view.layoutIfNeeded()
        }
        
        if (animated) {
            UIView.animateWithDuration(0.5,
                delay: 0.0,
                usingSpringWithDamping: 0.6,
                initialSpringVelocity: 8.0,
                options: UIViewAnimationOptions(rawValue: 0),
                animations: f,
                completion: nil)
        }
        else {
            f()
        }
    }
    
    private func closeDetails(animated: Bool) {
        let f: () -> () = {
            self.drawerTopConstraint.constant = -self.stationDetails.frame.size.height
            self.view.layoutIfNeeded()
        }
        
        if (animated) {
            UIView.animateWithDuration(0.5,
                delay: 0.0,
                usingSpringWithDamping: 0.6,
                initialSpringVelocity: -8.0,
                options: UIViewAnimationOptions(rawValue: 0),
                animations: f,
                completion: nil)
        }
        else {
            f()
        }
    }

    private var openedStation: TBStation? {
        return self.mapView.selectedAnnotations.first as? TBStation
    }
    
    private func availabilityTitleForStation(station: TBStation) -> String? {
        switch station.state {
        case .StationFull: return NSLocalizedString("No parking", comment: "")
        case .StationEmpty: return NSLocalizedString("No bicycles", comment: "")
        case .StationMarginal: return NSLocalizedString("Almost empty", comment: "")
        case .StationMarginalFull: return NSLocalizedString("Almost full", comment: "")
        case .StationInactive: return NSLocalizedString("Not operational", comment: "")
        case .StationUnknown: fallthrough
        case .StationOK: fallthrough
        default:
            return nil
        }
    }
    
    private func updateStationDetails(station: TBStation?, animated: Bool) {
        if station == nil {
            self.currentStationObserver = nil
            self.closeDetails(animated)
            self.updateTitle(nil)
            return
        }
        
        self.currentStationObserver = TBObserver.observerForObject(station, keyPath: "lastUpdateTime") {
            self.stationAvailabilityView.station = station
            let labelText = self.availabilityTitleForStation(station!)
            self.availabilityLabel.hidden = labelText == nil
            self.availabilityLabel.text = labelText
            self.availabilityLabel.textColor = station!.indicatorColor
            self.labelBackgroundView.hidden = self.availabilityLabel.hidden
        }
        
        self.openDetails(true)
        self.updateFavoriteButton(station!)
        self.updateTitle(station!.stationName)
    }
    
    private func updateFavoriteButton(station: TBStation) {
        self.toggleStationFavoriteButton.image = station.isFavorite ?
            UIImage(named: "station-favorite-selected") :
            UIImage(named: "station-favorite-unselected")
    }
    
    @IBAction private func toggleStationFavorite(sender: AnyObject!) {
        if let station = self.openedStation {
            
            // show a one-off alert with information on how favorites work
            if NSUserDefaults.standardUserDefaults().oneOff("favorites_alert_one_off") {
                TBAlerts.showAlertFromViewController(self,
                    title: NSLocalizedString("Star/Unstar Station", comment: ""),
                    message: NSLocalizedString("This station has been added to your list of favorites. Tap again to unstar", comment: ""))
            }
            
            station.favorite = !station.favorite
            self.updateFavoriteButton(station)
        }
    }
    
    @IBAction private func sendStationReport(sender: AnyObject!) {
        // gracefully bypass if city was not loaded and we don't have an email address
        if TBServer.instance.city == nil {
            return
        }
        
        if let station = self.openedStation {
            let openMailComposer: () -> () = {
                let vc = TBFeedbackMailComposeViewController(feedbackOption: TBFeedbackActionSheetService)
                vc.mailComposeDelegate = self
                vc.setSubject(NSLocalizedString("Problem in station \(station.sid)", comment: ""))
                let stationAddress = station.address ?? NSLocalizedString("N/A", comment: "")
                vc.setMessageBody(NSLocalizedString("Please describe the problem:\n\n\n=====================\nStation ID: \(station.sid)\nName: \(station.stationName)\nAddress: \(stationAddress)", comment: ""), isHTML: false)
                self.presentViewController(vc, animated: true, completion: nil)
            }
            
            if NSUserDefaults.standardUserDefaults().oneOff("report_oneoff") {
                TBAlerts.showAlertFromViewController(self,
                    title: NSLocalizedString("Contact Customer Service", comment: ""),
                    message: NSLocalizedString("An email addressed to Telofun customer service will be opened so you can report any issues with stations or individual bicycles", comment: ""),
                    dismissed: openMailComposer)
            }
            else {
                openMailComposer()
            }
        }
    }
    
    func mailComposeController(controller: MFMailComposeViewController, didFinishWithResult result: MFMailComposeResult, error: NSError?) {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    @IBAction private func navigateToStation(sender: AnyObject!) {
        if let station = self.openedStation {
            let dest = "\(station.coordinate.latitude),\(station.coordinate.longitude)"
            if !TBGoogleMapsRouting.routeFromAddress("", toAddress: dest) {
                TBAlerts.showAlertFromViewController(self, title: NSLocalizedString("Google Maps is not installed", comment: ""))
            }
        }
    }
    
    // - MARK: My Location
    
    func showMyLocation() {
        self.mapView.setUserTrackingMode(.Follow, animated: true)
    }

    func mapView(mapView: MKMapView, didFailToLocateUserWithError error: NSError) {
        print("ERROR: unable to determine user location: \(error)")
    }
    
    // - MARK: Bicycle Routes
    
    private var routesVisible: Bool = true
    private var kmlParser: KMLParser?

    private func showOrHideRoutesOnMap() {
        var showRoutes = true

        if let showBicycleRoutesValue = NSUserDefaults.standardUserDefaults().objectForKey("show_bicycle_routes") as! NSNumber? {
            showRoutes = showBicycleRoutesValue.boolValue
        }
        
        if showRoutes && !self.routesVisible {
            if self.kmlParser == nil {
                self.kmlParser = KMLParser(URL: NSBundle.mainBundle().URLForResource("routes", withExtension: "kml"))
                self.kmlParser?.parseKML()
            }
            
            self.mapView.addOverlays(self.kmlParser!.overlays as! [MKOverlay])
            self.routesVisible = true
            return
        }
        
        if !showRoutes && self.routesVisible {
            if let kmlParser = self.kmlParser {
                self.mapView.removeOverlays(kmlParser.overlays as! [MKOverlay])
            }
            self.routesVisible = false
        }
    }
    
    func mapView(mapView: MKMapView, rendererForOverlay overlay: MKOverlay) -> MKOverlayRenderer {
        return (self.kmlParser?.rendererForOverlay(overlay))!
    }
}